### ISSUE TITLE
Implementing CAPTCHA

### ISSUE THREAD

**Issue by nrbnlulu** [nrbnlulu/strawberry-django-auth#581]
### Discussed in https://github.com/nrbnlulu/strawberry-django-auth/discussions/580

<div type='discussions-op-text'>

<sup>Originally posted by **pkrakesh** October  1, 2024</sup>
How can I implement CloudFlare Turnstile or Google reCaptcha wth the package? These widget returns a token and in the backend we can verify the the token against its api endpoint. This is the way I have implemented the verification:

```
    @strawberry_django.input_mutation(handle_django_errors=True)
    def verify_human(self, info, token: str) -> CloudFlareTurnstileNode:
        request = info.context['request'] 
        url = "https://challenges.cloudflare.com/turnstile/v0/siteverify"
        ip = request.headers.get('CF_Connecting-IP')

        data = {
            "secret": settings.CLOUDFLARE_SECRET_KEY,
            "response": token,
            "ip": ip
        }

        response = requests.post(url, data=data)
        print(response.json())
        is_success = response.json()['success']
        return CloudFlareTurnstileNode(success=is_success)
```
```
# types.py

@strawberry.type
class CloudFlareTurnstileNode:
    success: bool
```
How do I add the verification to the `register` field provided by the package?</div>

**Comment by nrbnlulu** [comment]
Not sure it is currently possible in a proper way. we do have our own captcha implementation here though.

**Comment by pkrakesh** [comment]
I don't know how to implement the built in captcha as the documentation is not provided with much information. Can I get some help?

**Comment by pkrakesh** [comment]
The simple scenario is this as I understand:  

Frontend:

1. Adding a 'reCaptcha' widget in the frontend with with a `SITE_KEY`.
2. The widget generates a token as the user clicks "I'm not a Robot button".

Backend:

1. Send a POST request to the API endpoint of the widget with the `SECRET_KEY` and the token received from the frontend.
2. The response JSON contains a boolean flag `success`.
3. If `success == True`: "Good to go"

This is the way I have implemented the Google reCaptcha when I was using the Ariadne GraphQL library.
```
def resolve_create_user_mutation(_, info, input):
    email = input["email"]
    password = input["password"]
    recaptchaToken = input["recaptchaToken"]
    validation = validateHuman(recaptchaToken)

    if validation:
        user = User(username=email, email=email, currency=currency)
        user.set_password(password)
        user.save()
```
and I have implemented a custom method to validate the token against the widget's API endpoint:

```
def validateHuman(token):
    url = "https://www.google.com/recaptcha/api/siteverify"
    data = {"secret": settings.RECAPTCHA_SECRET_KEY, "response": SITE_KEY}
    response = requests.post(url, data=data)
    return response.json()["success"]
```
Is there any possibility to allow users to add a custom function like this for validating the token? Then we can implement any custom widget whether it is Google reCaptcha or Cloudflare Turnstile.

**Comment by nrbnlulu** [comment]
> I don't know how to implement the built in captcha as the documentation is not provided with much information. Can I get some help?

You don;t need to do much
just 
`LOGIN_REQUIRE_CAPTCHA=True`
`REGISTER_REQUIRE_CAPTCHA=True`

**Comment by pkrakesh** [comment]
Thank you. If there any possibility of Implementing Google the captcha supports as in this issue, it will be great as these are the most commonly used, modern and frictionless.

**Comment by nrbnlulu** [comment]
yeah sure we can add support for this. if you want to submit a PR go for it.
Otherwise in the future I plan to refactor the whole package to be with better extensibility so that 
users can easily support this themselves.

do note that in terms of security it doesn't really matter if you use reCaptcha or the internal captcha since AI can solve
pretty much everything these days...    https://chromewebstore.google.com/detail/rektcaptcha-recaptcha-sol/bbdhfoclddncoaomddgkaaphcnddbpdh?pli=1

**Comment by pkrakesh** [comment]
Refactoring is a good idea. One of the problems I have experienced with this package is the customization options. Currently, the only way we can achieve this by setting a variable in GQL_AUTH. It would be great if the library is less opinionated and extensible and as you have in mind. Currently this is the best for JWT Authentication for Strawberry Django.

**Comment by maxclax** [comment]
```
@strawberry.input
class CustomRegisterInput:
    email: str
    password1: str
    password2: str
    captcha: str


class CustomRegister(arg_mutations.Register):
    @strawberry.input
    class RegisterInput(types.CustomRegisterInput): ...

    @classmethod
    def resolve_mutation(
        cls, info, input_:CustomRegisterInput
    ) -> MutationNormalOutput:
        if hasattr(input_, "captcha"):
            # Validate the captcha
            captcha.captcha_token_validate(
                info.context.request, input_.captcha, raise_exception=True
            )

        return super().resolve_mutation(info, input_)



@strawberry.type
class AuthMutations:
    # --------------- Authentication mutations ---------------
    register = CustomRegister.field

```
